service/storage_service.cc

```diff
@@ -1681,13 +1682,16 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     if (_raft_experimental_topology) {
         auto holder = _async_gate.hold();
         // Waited on during stop()
-        (void)([] (storage_service& me, sharded<db::system_distributed_keyspace>& sys_dist_ks, gate::holder holder) -> future<> {
+        (void)([] (storage_service& me, sharded<db::system_distributed_keyspace>& sys_dist_ks, shared_ptr<i_endpoint_state_change_subscriber> gossip_feature_enabler, gate::holder holder) -> future<> {
             try {
                 co_await me.track_upgrade_progress_to_topology_coordinator(sys_dist_ks);
+                if (gossip_feature_enabler) {
+                    co_await me._gossiper.unregister_(gossip_feature_enabler);
```

Comment:
Isn't there a race here? Looks like the running of feature enabled overlaps a bit with the topology coordinator. So it may call enable_features() once or twice before we disable it. Not sure if it's a problem. Perhaps a correct solution would have enable_features() itself check where in the upgrade we are

Reply:
I can change it, thoguh I don't think the current code will be a problem. Features that are initially put into `system.topology` are completely the same as the features that are currently gossiped. `enable_features` will overwrite some state that is replicated from raft, but it will overwrite it with the same value (enabled features in `system.scylla_local`).

Reply:
Imagine a day passes between these two lines. Is it still correct? What if, let's say, topology coordinator enables some new feature during that time (not sure if possible)?

Reply:
> not sure if possible

That would require doing a rolling upgrade to a new version with new features, and after restart this code path wouldn't be touched at all.

Anyway, I get the point, we don't want to try to look at this code every time and have convince ourselves that it works. I will change the feature enabler to check the flag itself.